### PR TITLE
Added routing exporter

### PR DESCRIPTION
**Pull request by jpkrohling**
This PR depends on #1609, or a similar solution that will deliver all available exporters as part of `host.GetExporters`.

While I believe this is still in draft mode, I would like to validate the direction this is going.

Open points to discuss:
- [ ] what to do when an exporter fails?
- [ ] I left `processors` out for now, as it's probably more complex than exporters, given that there are no `host.GetProcessors`
- [x] I still need to do a complete e2e manual test
- [ ] this includes only traces for now. Once the idea is validated, I'll implement the same for the other exporter types.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
Manually tested with the following config:

```yaml
receivers:
  otlp/acme:
    protocols:
      grpc:
        endpoint: localhost:55680
  jaeger:
    protocols:
      thrift_compact:

processors:

exporters:
  logging: {}

  jaeger/default:
    endpoint: localhost:24250
    insecure: true

  otlp/acme:
    endpoint: localhost:55680
    insecure: true
    headers:
      "x-tenant": "acme"
  jaeger/acme:
    endpoint: localhost:14250
    insecure: true

  routing:
    from_attribute: X-Tenant
    default_exporters: jaeger/default
    table:
    - value: acme
      exporters: [jaeger/acme]

service:
  pipelines:
    traces:
      receivers: [jaeger]
      processors: []
      exporters: [logging, otlp/acme]

    traces/multitenant:
      receivers: [otlp/acme]
      processors: []
      exporters: [logging, routing]

    traces/routing:
      exporters:
      - jaeger/default
      - jaeger/acme
```

The `jaeger-acme` instance was started on my laptop (bare metal) as: 

```
go run ./cmd/all-in-one --query.static-files ./jaeger-ui/packages/jaeger-ui/build --processor.jaeger-compact.server-host-port :16831 --log-level=debug
```

The `jaeger-default` instance was started in a container, as:

```
podman run -p 24250:14250 -p 26686:16686 jaegertracing/all-in-one --log-level=debug
```

When the `x-tenant` header in the `otlp/acme` exporter is set to `acme`, the traces arrive correctly at `jaeger/acme` (bare metal). When the `x-tenant` is set to `globex`, it arrives at `jaeger/default` (container).

**Comment by jpkrohling** [comment]
As @tigrannajaryan and @jrcamp might be too busy, would you be able to review this one, @bogdandrutu, @dmitryax ?

File: exporter/routingexporter/routing.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	"go.uber.org/zap"
+	"google.golang.org/grpc/metadata"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const (
+	defaultKey = "__default"
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply exporter/routingexporter/routing.go:32]: Sounds reasonable.

**Comment by jpkrohling** [comment]
As this depends on #1609, this PR is blocked waiting on a resolution for that. Once a decision there is made, I'll work on the review comments and on the pending items from this PR's description.

**Comment by jpkrohling** [comment]
> I would also add this component to the contrib repo for now. If there is evidence that this is essential in the Collector we can then move it to the core repo.

Isn't there a discussion/plan on moving contrib to this main repo?

**Comment by tigrannajaryan** [comment]
> Isn't there a discussion/plan on moving contrib to this main repo?

No definitive plan yet.

**Comment by jpkrohling** [comment]
@tigrannajaryan, @nilebox, I rebased this PR and changed it to be a processor instead of exporter, as discussed in #1609.

**Comment by jpkrohling** [comment]
For reference, here's how to perform a manual test for this feature:

Configuration
```yaml
receivers:
  otlp/acme:
    protocols:
      grpc:
        endpoint: localhost:55680
  jaeger:
    protocols:
      thrift_compact:

processors:
  routing:
    from_attribute: X-Tenant
    default_exporters: jaeger/default
    table:
    - value: acme
      exporters:
      - jaeger/acme
  batch:

exporters:
  logging: {}

  jaeger/default:
    endpoint: localhost:24250
    insecure: true

  otlp/acme:
    endpoint: localhost:55680
    insecure: true
    headers:
      "x-tenant": "acme"
  jaeger/acme:
    endpoint: localhost:14250
    insecure: true

service:
  pipelines:
    traces:
      receivers:
      - jaeger
      processors: []
      exporters:
      - logging
      - otlp/acme

    traces/multitenant:
      receivers: [otlp/acme]
      processors: 
      - routing
      exporters:
      - logging
      - jaeger/default
      - jaeger/acme
```

Then, start a local Jaeger server (referenced as "jaeger/acme" in the config):
```console
$ go run ./cmd/all-in-one --processor.jaeger-compact.server-host-port=16831 --query.static-files=./jaeger-ui/packages/jaeger-ui/build --log-level=debug
```

And a second instance, from a container ("jaeger/default" in the config):

```console
$ podman run -p 24250:14250 -p 26686:16686 jaegertracing/all-in-one --log-level=debug
```

Jaeger's `tracegen` can be used to generate traces:
```console
$ go run ./cmd/tracegen -traces 10
```

Check "jaeger/acme" (http://localhost:16686), and you'll see the 10 traces there.

Change the configuration, so that the line with "x-tenant" has "globex" as value and reload otelcol. Run `tracegen` again, and check that the new traces arrived at the "jaeger/default" instance (http://localhost:26686). 

In a production setup, each tenant would have its own otelcol instance, with the appropriate `x-tenant` value.

**Comment by jpkrohling** [comment]
The CI failed at `correctness`, but I have no idea what's going on there and I don't think it's related to this PR:

```
    --- FAIL: TestTracingGoldenData/zipkin-zipkin (2.46s)
FAIL
exit status 1
FAIL	go.opentelemetry.io/collector/testbed/correctness/traces	137.314s
# Test Results
Started: Tue, 01 Sep 2020 11:14:02 +0000

Test                                    |Result|Duration|Sent Items|Received Items|Failure Count|Failures
----------------------------------------|------|-------:|---------:|-------------:|------------:|--------
TracingGoldenData/otlp-jaeger           |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/zipkin-opencensus     |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/otlp-opencensus       |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/jaeger-opencensus     |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/opencensus-jaeger     |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/zipkin-otlp           |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/jaeger-jaeger         |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/opencensus-opencensus |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/otlp-zipkin           |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/jaeger-zipkin         |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/opencensus-zipkin     |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/zipkin-jaeger         |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/otlp-otlp             |PASS  |    120s|      3907|          3907|            0|
TracingGoldenData/jaeger-otlp           |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/opencensus-otlp       |PASS  |      1s|      3907|          3907|            0|
TracingGoldenData/zipkin-zipkin         |FAIL  |      2s|      2529|          2529|            0|
```

https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector/3206/workflows/a7d8172d-1fe6-4576-b4b3-1d3baaf91ca5/jobs/31894

**Comment by jpkrohling** [comment]
And the CI error is now gone O_O

File: processor/routingprocessor/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Routing processor
+
+Routes traces to specific exporters.
+
+This processor will read a header from the incoming HTTP request (gRPC or plain HTTP) and direct the trace information to specific exporters based on the attribute's value.
+
+This processor *does not* let traces to continue through the pipeline and will emit a warning in case other processor(s) are defined after this one. Similarly, exporters defined as part of the pipeline are not authoritative: if you add an exporter to the pipeline, make sure you add it to this processor *as well*, otherwise it won't be used at all. All exporters defined as part of this processor *must also* be defined as part of the pipeline's exporters.
+
```
Comment (tigrannajaryan) [inline processor/routingprocessor/README.md:8]: We also need to tell that this processor cannot be used with `batch` and others which do not keep the incoming context, right?

**Comment by jpkrohling** [comment]
Closed in favor of open-telemetry/opentelemetry-collector-contrib#907

File: processor/routingprocessor/README.md (hunk 2 above)
Comment (jpkrohling) [reply processor/routingprocessor/README.md:8]: Done!

**Pull request closed without merging**